client/consensus/common/src/lib.rs

```diff
@@ -69,46 +81,173 @@ impl<B: BlockT> ParachainConsensus<B> for Box<dyn ParachainConsensus<B> + Send +
 	}
 }
 
+/// Value good enough to be used with parachains using the current backend implementation
+/// that ships with Substrate. This value may change in the future.
+pub const MAX_LEAVES_PER_LEVEL_SENSIBLE_DEFAULT: usize = 32;
+
+/// Upper bound to the number of leaves allowed for each level of the blockchain.
+///
+/// If the limit is set and more leaves are detected on block import, then the older ones are
+/// dropped to make space for the fresh blocks.
+///
+/// In environments where blocks confirmations from the relay chain may be "slow", then
+/// setting an upper bound helps keeping the chain health by dropping old (presumably) stale
+/// leaves and prevents discarding new blocks because we've reached the backend max value.
+pub enum LeavesLevelLimit {
+	/// Limit set to `MAX_LEAVES_PER_LEVEL_SENSIBLE_DEFAULT`.
+	Default,
+	/// No explicit limit, however a limit may be implicitly imposed by the backend implementation.
+	None,
+	/// Custom value.
+	Some(usize),
+}
+
+type BlockHash<Block> = <<Block as BlockT>::Header as HeaderT>::Hash;
+
+// Support structure to constrain the number of leaves for each level.
+struct LeavesLevelMonitor<Block: BlockT, BE> {
+	// Max number of leaves for each level.
+	level_limit: usize,
+	// Monotonic counter used to keep track of block import age (bigger is younger).
+	import_counter: u64,
+	// Map between block hash and age.
+	leaves_cache: HashMap<BlockHash<Block>, u64>,
+	// Backend reference to remove leaves on level saturation.
+	backend: Arc<BE>,
+}
+
+// Threshold after which we are going to cleanup our internal leaves cache by removing hashes that
+// doesn't belong to leaf blocks anymore.
+// This is a farly arbitrary value that can be changed in the future without breaking anything.
+const CLEANUP_THRESHOLD: usize = 64;
```

Comment:
In order to prevent adding another trait bound to get the finality notifications I opted out for a simpler and cheaper solution that in the end gives the same net result.

I just check if the last finalized block is changed since the last cleanup, if it is, I proceed to the cleanup.

Since in normal circumstances the finalization proceeds approximately at the same pace as the block authoring/import. I also opted out to don't attempt the cleanup on every imported block. But to check if there is something to cleanup every `CLEANUP_THRESHOLD` imported blocks.
(**on averaege** I prefer to cleanup a bunch of information in one shot instead of cleaning one block at a time)

Overall this prevents passing an extra param or adding an extra bound for a type implementing `BlockchainEvents` (that propagates all the way up as the `Blockchain` generic I've added).